### DISCUSSION TITLE
Can I save a PHPMailer object in a database after sending failed?

### DISCUSSION THREAD

**Discussion by JayEn84** [PHPMailer/PHPMailer#2476] [Q&A] [answered]
Not necessarily PHPMailer related question, but maybe interesting for other users of this fine library:

In my company, my colleagues work with lots of different HTML forms which send emails after submitting them. From time to time (let's say every 5,000th email) cannot be sent, because the hoster's smpt server is unavailable. I receive the error message from the PHPMailer object, which is fine, but in such cases I'd like to save those emails (including attachments) and store them in a database, so that I can try to resend them with a cron job. Is that possible in an easy way, like "exporting a PHPMailer object to blob" or something like that?
If that's not possible, I'd create a little more complex database table with lots of columns that makes it possible for me to rebuild the original email object. However, before I do all that work, I'd like to ask you guys if you have a better / an easier solution. :-)

**Comment by Synchro** [comment, marked as answer]
No, PHPMailer doesn't have a built-in means of doing this (other than stock serialize/unserialize) so you will need to reconstruct messages. Before doing that though, I'd look into making better use of your own mail server. Mail servers are really good at retrying, far more efficient than anything you could write in PHP!

**Reply by JayEn84** [reply]
Thanks for the hint! I'll ask the server provider for further information. Although I guess that initiating something like a "retry" is not possible when a connection to the mail server cannot be opened:

`Can not authenticate to IMAP server: [CLOSED] IMAP connection broken (authenticate) Fehlgeschlagen: Verbindung zu {[...]:993/imap/ssl/novalidate-cert}INBOX[...]`

**Reply by Synchro** [reply]
That's an IMAP connection that doesn't have much to do with actually sending messages, and also not something that PHPMailer does? Are you doing an upload-after-send over IMAP, like the gmail example shows? If so, that's a separate problem.

**Reply by JayEn84** [reply]
Oops, please excuse me, I copied a wrong log entry. :-D
Currently, I don't find the corresponding error in our logs, but I'll share it once it shows up again.

**Reply by JayEn84** [reply]
Apart from that, I'm going to test the idea to serialize a PHPMailer object with attachments. If this works, I think I'll go with that solution as it appears to be much easier to program instead of having to re-build the object. I never tried to serialize an object, so thanks a lot for that!

**Reply by Synchro** [reply]
Well, regardless, handling SMTP connection errors is exactly what mail servers do. They have fast and efficient queuing systems with exponential backoff, automatic retries, so it's always a good idea to let it deal with that for you, especially since you're required to do essentially nothing to make it happen!

**Reply by Synchro** [reply]
Serialising with attachments could be tricky if the attachments are deleted between serialise and unserialise, so you may need to reattach those.
